src/analyses/base.ml

```diff
@@ -94,27 +94,35 @@ struct
   (*There surely is a better way, because this means that often the wrong one gets chosen*)
   module VarH = Hashtbl.Make(CilType.Varinfo)
   module VarMap = Map.Make(CilType.Varinfo)
-  let array_map = VarH.create 20
+  let array_map = ref (VarH.create 20)
+
+  type marshal = attributes VarMap.t VarH.t
+
+  let array_domain_annotation_enabled = lazy (GobConfig.get_bool "annotation.goblint_array_domain")
 
   let add_to_array_map fundec arguments =
-    let rec pointedArrayMap = function
-      | [] -> VarMap.empty
-      | (info,value)::xs ->
-        match value with
-        | `Address t when hasAttribute "goblint_array_domain" info.vattr ->
-          let possibleVars = PreValueDomain.AD.to_var_may t in
-          List.fold_left (fun map arr -> VarMap.add arr (info.vattr) map) (pointedArrayMap xs) @@ List.filter (fun info -> isArrayType info.vtype) possibleVars
-        | _ -> pointedArrayMap xs
-    in
-    match VarH.find_option array_map fundec.svar with
-    | Some _ -> () (*We already have something -> do not change it*)
-    | None -> VarH.add array_map fundec.svar (pointedArrayMap arguments)
+    if Lazy.force array_domain_annotation_enabled then
+      let rec pointedArrayMap = function
+        | [] -> VarMap.empty
+        | (info,value)::xs ->
+          match value with
+          | `Address t when hasAttribute "goblint_array_domain" info.vattr ->
+            let possibleVars = PreValueDomain.AD.to_var_may t in
+            List.fold_left (fun map arr -> VarMap.add arr (info.vattr) map) (pointedArrayMap xs) @@ List.filter (fun info -> isArrayType info.vtype) possibleVars
+          | _ -> pointedArrayMap xs
+      in
+      match VarH.find_option !array_map fundec.svar with
+      | Some _ -> () (*We already have something -> do not change it*)
+      | None -> VarH.add !array_map fundec.svar (pointedArrayMap arguments)
 
   let attributes_varinfo info fundec =
-    let map = VarH.find array_map fundec.svar in
-    match VarMap.find_opt info map with
-    | Some attr ->  Some (attr, typeAttrs (info.vtype)) (*if the function has a different domain for this array, use it*)
-    | None -> Some (info.vattr, typeAttrs (info.vtype))
+    if Lazy.force array_domain_annotation_enabled then
+      let map = VarH.find !array_map fundec.svar in
+      match VarMap.find_opt info map with
+      | Some attr ->  Some (attr, typeAttrs (info.vtype)) (*if the function has a different domain for this array, use it*)
+      | None -> Some (info.vattr, typeAttrs (info.vtype))
+    else
+      None
```

Comment:
Are you sure?

```suggestion
      Some (info.vattr, typeAttrs (info.vtype))
```

Reply:
At the only two call sites of `arttribute_varinfo`, its result is directly passed as the second argument to `project_val`. In `project_val`, this parameter is only used when ` "annotation.goblint_array_domain"` is `true`.

Reply:
Ok!